homeassistant/components/opentherm_gw/sensor.py

```diff
@@ -1,20 +1,621 @@
 """Support for OpenTherm Gateway sensors."""
 
-import logging
+from dataclasses import dataclass
 
-from homeassistant.components.sensor import ENTITY_ID_FORMAT, SensorEntity
+import pyotgw.vars as gw_vars
+
+from homeassistant.components.sensor import (
+    ENTITY_ID_FORMAT,
+    SensorDeviceClass,
+    SensorEntity,
+    SensorEntityDescription,
+    SensorStateClass,
+)
 from homeassistant.config_entries import ConfigEntry
-from homeassistant.const import CONF_ID
+from homeassistant.const import (
+    CONF_ID,
+    PERCENTAGE,
+    UnitOfPower,
+    UnitOfPressure,
+    UnitOfTemperature,
+    UnitOfTime,
+    UnitOfVolumeFlowRate,
+)
 from homeassistant.core import HomeAssistant, callback
-from homeassistant.helpers.device_registry import DeviceInfo
-from homeassistant.helpers.dispatcher import async_dispatcher_connect
 from homeassistant.helpers.entity import async_generate_entity_id
 from homeassistant.helpers.entity_platform import AddEntitiesCallback
 
-from . import DOMAIN
-from .const import DATA_GATEWAYS, DATA_OPENTHERM_GW, SENSOR_INFO, TRANSLATE_SOURCE
+from . import OpenThermGatewayDevice
+from .const import DATA_GATEWAYS, DATA_OPENTHERM_GW
+from .entity import OpenThermEntity, OpenThermEntityDescription
+
+SENSOR_FLOAT_SUGGESTED_DISPLAY_PRECISION = 1
+
+
+@dataclass(frozen=True, kw_only=True)
+class OpenThermSensorEntityDescription(
+    SensorEntityDescription, OpenThermEntityDescription
+):
+    """Describes opentherm_gw sensor entity."""
+
 
-_LOGGER = logging.getLogger(__name__)
+SENSOR_INFO: tuple[tuple[list[str], OpenThermSensorEntityDescription], ...] = (
+    (
+        [gw_vars.BOILER, gw_vars.THERMOSTAT],
+        OpenThermSensorEntityDescription(
+            key=gw_vars.DATA_CONTROL_SETPOINT,
+            friendly_name_format="Control Setpoint {}",
+            device_class=SensorDeviceClass.TEMPERATURE,
+            state_class=SensorStateClass.MEASUREMENT,
+            native_unit_of_measurement=UnitOfTemperature.CELSIUS,
+            suggested_display_precision=SENSOR_FLOAT_SUGGESTED_DISPLAY_PRECISION,
+        ),
+    ),
+    (
+        [gw_vars.BOILER, gw_vars.THERMOSTAT],
+        OpenThermSensorEntityDescription(
+            key=gw_vars.DATA_MASTER_MEMBERID,
+            friendly_name_format="Thermostat Member ID {}",
+        ),
+    ),
+    (
+        [gw_vars.BOILER, gw_vars.THERMOSTAT],
+        OpenThermSensorEntityDescription(
+            key=gw_vars.DATA_SLAVE_MEMBERID,
+            friendly_name_format="Boiler Member ID {}",
+        ),
+    ),
+    (
+        [gw_vars.BOILER, gw_vars.THERMOSTAT],
+        OpenThermSensorEntityDescription(
+            key=gw_vars.DATA_SLAVE_OEM_FAULT,
+            friendly_name_format="Boiler OEM Fault Code {}",
+        ),
+    ),
```

Comment:
What does this return?

Reply:
Boiler manufacturers can return a manufacturer-specific fault code between `0` and `255`, which will be reported in this sensor.

Reply:
Do we have a list of them? Can we do an attempt to collect all the possible values from users and then complete the manufacturer specific list and add names to them so they can be contextualized?

Reply:
As they are manufacturer-specific and hopefully rare I doubt that's feasible. It would have to be combined with the manufacturer info, basically requiring a separate list of codes to be populated for every manufacturer out there.